manticore/core/smtlib/expression.py

```diff
@@ -1,33 +1,117 @@
+""" Module for Symbolic Expression
+
+ConstraintSets are considered a factory for new symbolic variables of type:
+BoolVariable, BitvecVariable and ArrayVariable.
+
+Normal python operators are overloaded in each class, complex expressions trees
+are built operating over expression variables and constants
+
+    cs = ConstraintSet()
+    x = cs.new_bitvec(name="SOMEVARNAME", size=32)
+    y = x + 199
+    condition1 = y < 1000
+    condition1 = x > 0
+
+    cs.add( condition1 )
+    cs.add( condition2 )
+
+"""
+
 from functools import reduce
-from ...exceptions import SmtlibError
 import uuid
-
 import re
 import copy
-from typing import Union, Optional, Dict, List
+from typing import Union, Optional, Tuple, List
+
 
+def local_simplify(e):
+    from .visitors import simplify as visitor_simplify
 
-class ExpressionException(SmtlibError):
+    return visitor_simplify(e)
+
+
+class ExpressionError(Exception):
     """
     Expression exception
     """
 
     pass
 
 
-class Expression:
+class XSlotted(type):
+    """Metaclass that will propagate slots on multi-inheritance classes
+    Every class should define __xslots__ (instead of __slots__)
+
+    class Base(object, metaclass=XSlotted, abstract=True):
+        pass
+
+    class A(Base, abstract=True):
+        __xslots__ = ('a',)
+        pass
+
+    class B(Base, abstract=True):
+        __xslots__ = ('b',)
+        pass
+
+    class C(A, B):
+        pass
+
+    # Normal case / baseline
+    class X(object):
+        __slots__ = ('a', 'b')
+
+    c = C()
+    c.a = 1
+    c.b = 2
+
+    x = X()
+    x.a = 1
+    x.b = 2
+
+    import sys
+    print (sys.getsizeof(c),sys.getsizeof(x)) #same value
+
+    """
+
+    @staticmethod
+    def _remove_mod(attr: str) -> str:
+        """xlots attrivutes could have modifficators after a # symbol
+        attribute#v  means attribute is _volatile_ and should not be saved to storage
+        """
+        return attr.split("#")[0]
+
+    def __new__(cls, clsname, bases, attrs, abstract=False):
+
+        xslots = set(attrs.get("__xslots__", ()))
+        # merge the xslots of all the bases with the one defined here
+        for base in bases:
+            xslots = xslots.union(getattr(base, "__xslots__", ()))
+        attrs["__xslots__"]: Tuple[str] = tuple(xslots)
+        if abstract:
+            attrs["__slots__"] = ()
+        else:
+            attrs["__slots__"]: Tuple[str] = tuple(map(cls._remove_mod, attrs["__xslots__"]))
+
+        return super().__new__(cls, clsname, bases, attrs)
+
+
+class Expression(object, metaclass=XSlotted, abstract=True):
```

Comment:
It's been a while since I do not entertain that idea. Originally my concern was about it doing some slow dynamic checks.
I'd like to see how an ABC version of this may look and run.
We still would need a way for handle slots in multi-inherited class thing.